File: Medium/LC133- Clone-Graph/code.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraph(self, node):
        oldToNew = {}
        def clone(node):
            if node in oldToNew:
                return oldToNew[node]
            copy = Node(node.val)
            oldToNew[node] = copy
            for nd in node.neighbors:
                copy.neighbors.append(clone(nd))
            return copy
        return clone(node) if node else None

        

File: Medium/LC133- Clone-Graph/code.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    unordered_map<Node*, Node*> mp;
    Node* dfs(Node* node) {

        Node* copy = new Node(node->val);
        mp[node] = copy;
        
        for(auto nd : node->neighbors) {
            
            if(mp.find(nd) != mp.end()) {
                copy->neighbors.push_back(mp[nd]);
            }else {
                copy->neighbors.push_back(dfs(nd));
            }

        }
        return copy;
    }
    Node* cloneGraph(Node* node) {

        if(node == NULL) {
            return NULL;
        }
        return dfs(node);
    }
};
